modxcms/revolution
# [Port to 3.x] Fix Various Issues with Resource/Element Validation and Field Display

### What does it do?
Ports enhancements from PR #15146 to 3.x (including subsequent fixes by others) and incorporates UI updates introduced in PR #15773 made to the TV editing form into the rest of the element editing forms. Other changes include:
1. Consolidated repetitive method definitions found in the element widget classes, placing single replacement methods into the parent FormPanel class.
2. Addressed two concerns raised by @Ruslan-Aleev in #15854 : (1) Fine-tuned style for tag copy to make clear that the icon that shows on hover is only a hint and cannot be clicked to perform the copy, and (7) made style updates so only the settings-type checkboxes in the resource editing form display in the new switch style by default (TVs show the normal checkbox; a new input option setting could be created to control this by assigning the class I added in the forms SCSS ["display-switch"] in a future PR).
3. Made language changes to the element tab introduction text for clarity, including an anchor link to the element code (when applicable)

### Why is it needed?
Makes needed UI and consistency improvements to the resource and element editing panels.

### How to test

1. From your terminal app, run `grunt build` from within the `_build/templates/default` folder and clear your browser cache.
2. Create a handful of TVs (at least one checkbox TV and one TV of any type that is _required_) and verify they operate and appear as expected.
3. Move at least one _required_ TV via Form Customization and assign it to a template. From a resource using this template, verify that (when this TV is not filled and you try to save the resource while viewing a tab other than the one containing this TV) errored tabs are opened and navigated to on save.
4. Create at least one test element for each type (chunk, plugin, template, tv), and verify each appear, behave, and save as expected.

### Related issue(s)/PR(s)
Partly addresses #15854.
